pkg/innerring/subnet.go

```diff
@@ -0,0 +1,332 @@
+	s.registerStarter(func() error {
+		var err error
+
+		// initialize queue for processing of the events from Subnet contract
+		s.subnetHandler.workerPool, err = ants.NewPool(int(cfg.queueSize), ants.WithNonblocking(true))
+		if err != nil {
+			return fmt.Errorf("subnet queue initialization: %w", err)
+		}
+
+		// initialize morph client of Subnet contract
+		clientMode := morphsubnet.NotaryAlphabet
+
+		if s.sideNotaryConfig.disabled {
+			clientMode = morphsubnet.NonNotary
+		}
+
+		var initPrm morphsubnet.InitPrm
+
+		initPrm.SetBaseClient(s.morphClient)
+		initPrm.SetContractAddress(s.contracts.subnet)
+		initPrm.SetMode(clientMode)
+
+		err = s.subnetHandler.morphClient.Init(initPrm)
+		if err != nil {
+			return fmt.Errorf("init morph subnet client: %w", err)
+		}
+
+		s.listenSubnet()
+
+		return nil
+	})
+
+	s.registerCloser(func() error {
+		s.stopSubnet()
+		return nil
+	})
+}
+
+// releases the Subnet contract notification processing queue.
+func (s *Server) stopSubnet() {
+	s.workerPool.Release()
+}
+
+// names of listened notification events from Subnet contract.
+const (
+	// subnet creation
+	subnetCreateEvName = "put"
+	// subnet removal
+	subnetRemoveEvName = "delete"
+)
+
+// makes IR server to listen notifications of Subnet contract.
+// All required resources must be initialized before (initSubnet).
+// Works in one of two modes (configured): notary and non-notary.
+//
+// All handlers are executed only if local node is an alphabet one.
+//
+// Events (notary):
+//   * put (parser: subnetevents.ParseNotaryPut, handler: catchSubnetCreation);
+//   * delete (parser: subnetevents.ParseNotaryDelete, handler: catchSubnetRemoval).
+//
+// Events (non-notary):
+//   * put (parser: subnetevents.ParsePut, handler: catchSubnetCreation);
+//   * delete (parser: subnetevents.ParseDelete, handler: catchSubnetCreation).
+func (s *Server) listenSubnet() {
+	if s.sideNotaryConfig.disabled {
+		s.listenSubnetWithoutNotary()
+		return
+	}
+
+	var (
+		parserInfo  event.NotaryParserInfo
+		handlerInfo event.NotaryHandlerInfo
+	)
+
+	parserInfo.SetScriptHash(s.contracts.subnet)
+	handlerInfo.SetScriptHash(s.contracts.subnet)
+
+	listenEvent := func(notifyName string, parser event.NotaryParser, handler event.Handler) {
+		notifyTyp := event.NotaryTypeFromString(notifyName)
+
+		parserInfo.SetMempoolType(mempoolevent.TransactionAdded)
+		handlerInfo.SetMempoolType(mempoolevent.TransactionAdded)
+
+		parserInfo.SetParser(parser)
+		handlerInfo.SetHandler(handler)
+
+		parserInfo.SetRequestType(notifyTyp)
+		handlerInfo.SetRequestType(notifyTyp)
+
+		s.morphListener.SetNotaryParser(parserInfo)
+		s.morphListener.RegisterNotaryHandler(handlerInfo)
+	}
+
+	// subnet creation
+	listenEvent(subnetCreateEvName, subnetevents.ParseNotaryPut, s.onlyAlphabetEventHandler(s.catchSubnetCreation))
+	// subnet removal
+	listenEvent(subnetRemoveEvName, subnetevents.ParseNotaryDelete, s.onlyAlphabetEventHandler(s.catchSubnetRemoval))
```

Comment:
in notary enabled environmental you need to listen to just `Delete` notifications